core-primitives/stf-executor/src/lib.rs

```diff
@@ -102,23 +103,39 @@ impl ExecutedOperation {
 ///
 /// Contains multiple executed operations
 #[derive(Clone, Debug)]
-pub struct BatchExecutionResult {
-	pub initial_state_hash: H256,
+pub struct BatchExecutionResult<Externalities: SgxExternalitiesTrait + Encode> {
+	pub state_hash_before_execution: H256,
 	pub executed_operations: Vec<ExecutedOperation>,
+	pub state_after_execution: Externalities,
 }
 
-impl BatchExecutionResult {
+impl<Externalities> BatchExecutionResult<Externalities>
+where
+	Externalities: SgxExternalitiesTrait + Encode,
+{
 	pub fn get_extrinsic_callbacks(&self) -> Vec<OpaqueCall> {
 		self.executed_operations
 			.iter()
 			.flat_map(|e| e.status.get_extrinsic_callbacks())
 			.collect()
 	}
 
+	/// Returns all successfully exectued operation hashes.
 	pub fn get_executed_operation_hashes(&self) -> Vec<H256> {
 		self.executed_operations
 			.iter()
 			.flat_map(|ec| ec.status.get_executed_operation_hash())
 			.collect()
 	}
+
+	/// Returns all operations that were not executed.
+	pub fn get_failed_operations(&self) -> Vec<ExecutedOperation> {
+		self.executed_operations
+			.iter()
+			.flat_map(|ec| match ec.is_success() {
+				false => Some(ec.clone()),
+				true => None,
+			})
+			.collect()
+	}
```

Comment:
This calls for a nice unit test 😃 (I know you're working on that)

Reply:
Added these. And there will be a lot more incoming next week 🥳